.github/workflows/interop-test.yml

```diff
@@ -1,32 +1,37 @@
 name: Interoperability Testing
-
 on:
   pull_request:
   push:
     branches:
-      - master
-
-concurrency:
-  group: ${{ github.workflow }}-${{ github.ref }}
-  cancel-in-progress: true
+      - "master"
 
 jobs:
-  # NOTE: during a pull request run, github creates a merge commit referenced in `github.sha`
-  # that merge commit is not a regular commit. You won't find it with a regular `git checkout SHA` and
-  # tools like `go get repo@SHA` won't find it.
-  #
-  # As a workaround, we generate a path to the actual pull request's commit, it looks like:
-  # `github.com/external-org/go-libp2p@latest-commit-on-their-branch`
-  run-ping-interop-cross-version:
-    uses: "libp2p/test-plans/.github/workflows/run-composition.yml@master"
-    with:
-      composition_file: "ping/_compositions/rust-cross-versions.toml"
-      custom_git_target: github.com/${{ github.event.pull_request.head.repo.full_name || github.event.repository.full_name }}
-      custom_git_reference: ${{ github.event.pull_request.head.sha || github.sha }}
-  run-ping-interop-cross-implementation:
-    uses: "libp2p/test-plans/.github/workflows/run-composition.yml@master"
+  build-ping-container:
+    name: Build Ping interop container
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v3
+        with:
+          submodules: recursive
+      - name: Build image
+        working-directory: ./test-plans
+        run: make
+      - name: Upload ping versions info
+        uses: actions/upload-artifact@v3
+        with:
+          name: ping-versions
+          path: ./test-plans/ping-versions.json
+      - name: Upload image tar
+        uses: actions/upload-artifact@v3
+        with:
+          name: ping-image
+          path: ./test-plans/ping-image.tar
+  run-multidim-interop:
+    name: Run multidimensional interoperability tests
+    needs: build-ping-container
+    uses: "libp2p/test-plans/.github/workflows/run-testplans.yml@master"
     with:
-      composition_file: "ping/_compositions/all-interop-latest.toml"
-      custom_git_target: github.com/${{ github.event.pull_request.head.repo.full_name || github.event.repository.full_name }}
-      custom_git_reference: ${{ github.event.pull_request.head.sha || github.sha }}
-      custom_interop_target: rust
+      dir: "multidim-interop"
+      extra-versions: ping-versions
```

Comment:
What does `extra-versions` here mean? @MarcoPolo

We seem to be referencing the same image in there that we are also tarring up.

Reply:
The source of truth for the capabilities of a released libp2p version are defined in [`version.ts`](https://github.com/libp2p/test-plans/blob/master/multidim-interop/versions.ts). The extra versions is more version data that is added to that `versions.ts` array. This lets you add a "rust-libp2p-head" version to the defined versions and run an interop test between all defined versions on all capabilites. This combined with the ability to filter tests lets you run the interop tests that only include "rust-libp2p-head".

Reply:
Thanks for explaining, this makes sense now.

Is there a usecase where we extend `versions.ts` with more than one image and/or don't filter by the one we just added? It looks like the configuration API of this action could be simpler, i.e. it just needs the contents of one entry in `versions.ts` and we always want to filter the tests for this one version.

Reply:
In particular, this seems to be sufficient and perhaps a bit easier to understand:

```yaml
  run-multidim-interop:
    name: Run multidimensional interoperability tests
    needs: build-ping-container
    uses: "libp2p/test-plans/.github/workflows/run-testplans.yml@master"
    with:
      imageId: "rust-libp2p-head"
      transports: ["ws", "tcp", "quic-v1", "webrtc"]
      secureChannels: ["tls", "noise"]
      muxers: ["mplex", "yamux"]
```

- The name of the version can be implied from the image ID (by convention we should make this a tag and not just a hash).
- With the composite action suggested in https://github.com/libp2p/rust-libp2p/pull/3331/files#r1073550520, we don't need to pass the artifact name of the tar'd image.
- By directly specifying the transports, secure channels and muxers, we can avoid a configuration file in our repository.
- Not sure I understand why `dir: "multidim-interop"` needs to be specified, can't this be implied by the GitHub action?

Reply:
> Not sure I understand why dir: "multidim-interop" needs to be specified, can't this be implied by the GitHub action?

This should be changed. Originally the "run test-plans" was generic enough that it would work for any `npm test` like thing. But there's little value for this, and it would be nicer to be clearer here.

> By directly specifying the transports, secure channels and muxers, we can avoid a configuration file in our repository.

It's true. With the tradeoff of dealing not being able to run this locally as easily. I think it's nice to be able to locally run: `npm test -- --extra-versions=$RUST_LIBP2P/test-plans/version.json --name-filter="rust-libp2p-head"`. Inputs are also only strings, so you'd have to stringify the given example.

I agree with all the other points :)

Reply:
> > By directly specifying the transports, secure channels and muxers, we can avoid a configuration file in our repository.
> 
> It's true. With the tradeoff of dealing not being able to run this locally as easily. I think it's nice to be able to locally run: `npm test -- --extra-versions=$RUST_LIBP2P/test-plans/version.json --name-filter="rust-libp2p-head"`. Inputs are also only strings, so you'd have to stringify the given example.

How often do we need that? At the moment, this config file just sits there, not co-located with its consumer and it requires a fair amount of knowledge to figure out what it is for.

Reply:
> It's true. With the tradeoff of dealing not being able to run this locally as easily. I think it's nice to be able to locally run: `npm test -- --extra-versions=$RUST_LIBP2P/test-plans/version.json --name-filter="rust-libp2p-head"`. Inputs are also only strings, so you'd have to stringify the given example.

Would it make sense to unify the API of the test runner to just accept a series of JSON files with one version each? i.e. it builds the matrix of tests out of all files that are passed to it. Each JSON file would contain exactly one entry.

We could then pass the JSON as a string into the workflow and it writes it to a file. If you really need to run things locally, you just have to quickly copy paste that and make the file.

To me, that seems like an acceptable trade-off for having one less configuration file around per repository :)

Reply:
> Would it make sense to unify the API of the test runner to just accept a series of JSON files with one version each? i.e. it builds the matrix of tests out of all files that are passed to it. Each JSON file would contain exactly one entry.

Yeah, this seems reasonable 👍 .

> We could then pass the JSON as a string into the workflow and it writes it to a file. If you really need to run things locally, you just have to quickly copy paste that and make the file.

This less logic that is in the github actions yaml the better. At least for me, I find this stuff hard to work with since it only runs in CI. The simpler the yaml file is the better.

I don’t see why having this configuration file is bad. I actually think it’s a good thing because:
1. It defines what the capabilities of the current rust-libp2p are.
2. If a new capability is added (e.g. webtransport), then you modify this json file, not a stringified version in a github action yaml file.
3. When a release is made, you simply have to copy this JSON file to `test-plans/multidim-interop/version.ts` so other implementions test against it.

Reply:
> > Would it make sense to unify the API of the test runner to just accept a series of JSON files with one version each? i.e. it builds the matrix of tests out of all files that are passed to it. Each JSON file would contain exactly one entry.
> 
> Yeah, this seems reasonable 👍 .
> 

Nice, I'll open an issue on the repo.

> I don’t see why having this configuration file is bad. I actually think it’s a good thing because:
> 1. It defines what the capabilities of the current rust-libp2p are.
> 2. If a new capability is added (e.g. webtransport), then you modify this json file, not a stringified version in a github action yaml file.
> 3. When a release is made, you simply have to copy this JSON file to `test-plans/multidim-interop/version.ts` so other implementions test against it.

You are mentioning some good points. Happy to keep the file, thanks for getting to the bottom of this :)